tests/integration/framework/docker_runner.py

```diff
@@ -8,49 +8,93 @@
 import sys
 
 class DockerRunner:
-    """Run docker containers for testing"""
+    """Run a docker container in background to execute testing commands"""
 
     def __init__(self, platform, image, verbose):
         """Sets platform and image for all tests ran with this instance"""
         self.platform = platform
         self.image = image
         self.verbose = verbose
+        self.container_id = None
 
-    def construct_docker_command(self, envs, args):
+        # Launch the container
+        self._start()
+
+    def execute(self, envs, args):
+        """Launch `docker exec` commands inside the background container"""
+        command = self._construct_command("exec", envs, args)
+        return self._shell(command)
+
+    def run(self, envs, args):
+        """
+        Launch `docker run` commands to create a new container for each command
+        """
+        command = self._construct_command("run", envs, args)
+        return self._shell(command)
+
+    def __del__(self):
+        """Remove test container"""
+        stop_command = ["docker", "rm", "-f", self.container_id]
```

Comment:
If we want to retain the container (to investigate because an error happened) it's useful to have a 2-step process: `docker stop` the container, and then `docker rm` the container. We can then choose to not rm on test failure, allowing a `docker start` and perhaps `exec` into a shell to investigate.

Can we split this up into 2 commands that are called independent from the destructor?

Reply:
With our 2 type of command, `exec` & `run`:

- `run` commands outputs (stdout + stderr) are direclty handled. Stdout is returned, stderr displayed when error occurs (see `DockerRunner._shell`). Logs are already caught.
- `exec` commands do not add any logs, they seem to handle only main process outputs.

We can do this, but logs are already displayed, not sure where we would use it.